compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -809,18 +807,23 @@ public void visit(BLangImportPackage importPkgNode) {
         this.env.scope.define(pkgAlias, symbol);
     }
 
-    public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules, SymbolEnv env) {
+    public void initPredeclaredModules(Map<Name, BPackageSymbol> predeclaredModules,
+                                       List<BLangCompilationUnit> compUnits, SymbolEnv env) {
         SymbolEnv prevEnv = this.env;
         this.env = env;
-        for (Name alias : predeclaredModules.keySet()) {
-            ScopeEntry entry = this.env.scope.lookup(alias);
-            if (entry == NOT_FOUND_ENTRY) {
-                this.env.scope.define(alias, predeclaredModules.get(alias));
-            } else {
-                while (entry.next != NOT_FOUND_ENTRY) {
-                    entry = entry.next;
+        for (BLangCompilationUnit compUnit : compUnits) {
+            for (Name alias : predeclaredModules.keySet()) {
+                BPackageSymbol symbol = duplicatePackagSymbol(predeclaredModules.get(alias));
+                symbol.compUnit = new Name(compUnit.name);
+                ScopeEntry entry = this.env.scope.lookup(alias);
+                if (entry == NOT_FOUND_ENTRY) {
+                    this.env.scope.define(alias, symbol);
+                } else {
+                    while (entry.next != NOT_FOUND_ENTRY) {
+                        entry = entry.next;
+                    }
+                    entry.next = new ScopeEntry(symbol, NOT_FOUND_ENTRY);
                 }
-                entry.next = new ScopeEntry(predeclaredModules.get(alias), NOT_FOUND_ENTRY);
```

Comment:
How about changing this to something like

```java
                if (entry == NOT_FOUND_ENTRY) {
                    this.env.scope.define(alias, symbol);
                    continue;
                }

                while (entry.next != NOT_FOUND_ENTRY) {
                    entry = entry.next;
                }
                entry.next = new ScopeEntry(symbol, NOT_FOUND_ENTRY);
```

Reply:
Refactor `initPredeclaredModules` function as mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/26101#discussion_r505188467